airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_with_global.py

```diff
@@ -0,0 +1,179 @@
+    """
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: PartitionRouter, stream_cursor: DatetimeBasedCursor):
+        self._partition_router = partition_router
+        self._per_partition_cursor = PerPartitionCursor(cursor_factory, partition_router)
+        self._global_cursor = GlobalSubstreamCursor(stream_cursor, partition_router)
+        self._use_global_cursor = False
+
+    def _get_active_cursor(self) -> Union[PerPartitionCursor, GlobalSubstreamCursor]:
+        return self._global_cursor if self._use_global_cursor else self._per_partition_cursor
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        self._global_cursor.start_slices_generation()
+
+        # Iterate through partitions and process slices
+        for partition, is_last_partition in iterate_with_last_flag(self._partition_router.stream_slices()):
+            # Generate slices for the current cursor and handle the last slice using the flag
+            if partition is None:
+                continue
+            for slice, is_last_slice in iterate_with_last_flag(
+                self._get_active_cursor().generate_slices_from_partition(partition=partition)
+            ):
+                if slice is None:
+                    continue
+                self._global_cursor.register_slice(is_last_slice and is_last_partition)
+                yield slice
+
+    def set_initial_state(self, stream_state: StreamState) -> None:
+        """
+        Set the initial state for the cursors.
+        """
+        self._use_global_cursor = stream_state.get("use_global_cursor", False)
+
+        self._global_cursor.set_initial_state(stream_state)
+        self._per_partition_cursor.set_initial_state(stream_state)
+
+    def observe(self, stream_slice: StreamSlice, record: Record) -> None:
+        if not self._use_global_cursor and self._per_partition_cursor.limit_reached():
+            self._use_global_cursor = True
+
+        if not self._use_global_cursor:
+            self._per_partition_cursor.observe(stream_slice, record)
+        self._global_cursor.observe(stream_slice, record)
+
+    def close_slice(self, stream_slice: StreamSlice, *args: Any) -> None:
+        if not self._use_global_cursor:
+            self._per_partition_cursor.close_slice(stream_slice, *args)
+        self._global_cursor.close_slice(stream_slice, *args)
+
+    def get_stream_state(self) -> StreamState:
+        final_state = {"use_global_cursor": self._use_global_cursor}
+
+        final_state.update(self._global_cursor.get_stream_state())
+        if not self._use_global_cursor:
+            final_state.update(self._per_partition_cursor.get_stream_state())
+
+        return final_state
+
+    def select_state(self, stream_slice: Optional[StreamSlice] = None) -> Optional[StreamState]:
+        return self._get_active_cursor().select_state(stream_slice)
+
+    def get_request_params(
```

Comment:
Not something that needs to be addressed in this PR. But something worth noting, as part of the low-code concurrent work, we've effectively tried to decouple Cursor from RequestOptionProvider. See here for an example: https://github.com/airbytehq/airbyte/pull/45413/files#diff-a41404024e1da8bece6cd3c74f6a04294693b8c2edd59e5975a592ffb4e20888R32-R39

Because these request options provider methods rely on `_active_cursor` from the cursor to decide which params to provide, this might pose a problem when we want decouple the two concepts. Might be something we need to think about once we start trying to parallelize substreams. Not a blocker though